### DISCUSSION TITLE
Some ideas for future WinForms improvements

### DISCUSSION THREAD

**Discussion by noseratio** [dotnet/winforms#6184] [Ideas] [answered]
Bellow is a list of ideas for potential WinForms improvements I've come up with, while working closely with WinForms over the last few months. It was sparkled by [this Twitter thread](https://twitter.com/ericlaw/status/1459662971131604994?s=20) I participated in, where some folks suggested turning it into a repo discussion (cc @OliaG, @merriemcgaw, @RussKie, @KlausLoeffelmann), so here we go 🙂

- An option on `WindowsFormsSynchronizationContext` to prevent reentrancy issues caused by pumping while blocking. E.g., see how it may affect folks using SkiaSharp:
    
    https://github.com/mono/SkiaSharp/issues/1383#issuecomment-927750884

    (or unseal `WindowsFormsSynchronizationContext` to be subclass-able, [so we can overrride `SynchronizationContext.Wait`](https://github.com/mono/SkiaSharp/issues/1383#issuecomment-927307185))

- [inb4](https://github.com/dotnet/winforms/issues/4631), a proper `Control.InvokeAsync` implementation. E.g., see how the lack of it currently affects `BlazorWebView`:

    https://github.com/dotnet/maui/issues/2313

- `FlowLayoutPanel`'s children anchoring behavior ([they don't stretch in the way one may expect](https://docs.microsoft.com/en-us/dotnet/desktop/winforms/controls/how-to-anchor-and-dock-child-controls-in-a-flowlayoutpanel-control?view=netframeworkdesktop-4.8#:~:text=the%20first%20button.-,It%20does%20not%20stretch%20across%20the%20width%20of%20the%20FlowLayoutPanel%20control,-.%20This%20is%20the)). I don't understand why the child controls are stretched to the widest peer, while they are set to anchor or dock to the parent panel's right/left. I had to come up with a workaround that attaches a custom resizing behavior to each child control, which triggers in when the parent `FlowLayoutPanel` changes its size.

    In my opinion, `TableLayoutPanel`, `FlowLayoutPanel`, Docking, Anchoring and `AutoSize` are the most important building blocks for creating responsive UI in WinForms, so it'd be great to have an option to enable consistent sizing behavior across all of them.  

- `ShowDialogAsync` API: https://stackoverflow.com/a/33411037/1768303

- Work with the `WebView2` team to solve the pending keyboard accelerators and focus handling integration issues: https://github.com/noseratio/BlazorWebViewWinFormsApp.

    [This might be tricky](https://github.com/MicrosoftEdge/WebView2Feedback/issues/468#issuecomment-855810839) (given the out-of-proc nature of `WebView2`), but should be doable. I believe WinForms has the required input handling plumbing in place already (unlike WPF).

  A smooth integration with `WebView2` is particularly important. I'm sure we'll see more and more of hybrid UI desktop apps in the wild. WinForms is an ideal host for `WebView2` (just add a menu, status bar, tray icon and a WV2 control), but those are still the issues to resolve, in particular they may affect accessibility.

I'm happy to share what solutions/workarounds I currently use for the above (some are linked already), if there is any interest.

Thanks for considering it. We really appreciate that WinForms as a UI framework has always been actively maintained 💗🤞

**Comment by KlausLoeffelmann** [comment]
I am sorry that I am completely missed this.
Is this something which would be important for you?

As you can see, we already have a series of new API suggestions, we'd like to take on.
I would very much like to add discussions point there, and - if you're still up to that - create new API suggestion issue which we can more concretely talk about!

**Reply by noseratio** [reply]
Hi @KlausLoeffelmann, thanks for the follow-up! That feedback was based on a project I worked on at that time, which by now has been signed, sealed, delivered - using a homebrew version of those APIs.

I'll see what I can come up with in terms of the API proposals, but anything I might say now might not be be a fresh take, I'm afraid. As many others, I'm in between jobs nowadays and it doesn't look like my next role will be related to WinForms (as much as I enjoyed working with it!) or even to the same industry.

**Reply by noseratio** [reply]
While I still remember though... adding an option to create a WinForms app as a [.NET Generic Host app](https://learn.microsoft.com/en-us/dotnet/core/extensions/generic-host), with the main UI thread as one of `IHostedService` services, proper life-time management etc. might be a worthwhile proposal. I used this approach in that recent WinForms project of mine and I really liked the structure it imposes upon the whole app.

**Reply by RussKie** [reply]
> While I still remember though... adding an option to create a WinForms app as a [.NET Generic Host app](https://learn.microsoft.com/en-us/dotnet/core/extensions/generic-host), with the main UI thread as one of `IHostedService` services, proper life-time management etc. might be a worthwhile proposal. I used this approach in that recent WinForms project of mine and I really liked the structure it imposes upon the whole app.

Sounds similar to https://github.com/dotnet/winforms/issues/2861.

Do you a sample you could share?

**Reply by noseratio** [reply]
@RussKie I don't have a minimalistic sample to share right away, but I've just searched and it looks like other folks are doing a very similar thing: https://stackoverflow.com/a/60154498/1768303.

**Reply by noseratio** [reply]
Seems like that SO answer is actually coming from @alex-oswald, who proposed #2861.

**Reply by alex-oswald** [reply]
Updated repo page:
https://github.com/alex-oswald/WindowsFormsLifetime

**Reply by elachlan** [reply]
Can we reopen the linked issue?

**Reply by elachlan** [reply]
@alex-oswald are you okay if your WindowsFormsLifetime code is included into the Winforms Codebase is some form or another? It would obviously have to go through the API review, which I suspect the team has zero time for before .net 8 unfortunately.

@RussKie should I open a new issue based on #2861, or should we re-open that issue?

**Reply by RussKie** [reply]
Thank you for keeping the discussion alive. Personally, I'd love to see something like what @alex-oswald has done as a Windows Forms out-of-the-box feature. 
However, priority- or way-forward-wise, the question is for 
- @alex-oswald - whether Alex is willing to do a formal API proposal, go to the API Review and provide an implementation (something like #146 and #1133 come to mind), and 
- @merriemcgaw and the team - whether this feature is something the team would consider, and if so, the resource allocation to support Alex in this journey (e.g., something like #1133).

**Reply by alex-oswald** [reply]
@elachlan I am. I created #2861 with that intention, but at the time couldn't commit.

@RussKie I will happily commit to start working on a proposal and implementation if the team is willing to consider this addition.

**Comment by KlausLoeffelmann** [comment, marked as answer]
While I found this super-interesting, I am afraid we're currently missing the resources to prioritize this. But, let's discuss this in the next runtime meeting, therefore tagging @merriemcgaw, @JeremyKuhne, @dreddy-work, @Shyam-Gupta, @Tanya-Solyanik.

**Reply by elachlan** [reply]
#2861 got closed again. Are you able to reopen it and avoid the auto-close?

**Reply by merriemcgaw** [reply]
Yep. 😄
